impl/client/src/main/javascript/web/pentaho/type/Context.js

```diff
@@ -16,40 +16,278 @@
 define([
   "require",
   "module",
-  "../service",
   "../typeInfo",
   "../i18n!types",
-  "./standard",
+  "./InstanceConstructionContext",
   "./SpecificationContext",
   "./SpecificationScope",
+  "./InstancesContainer",
   "../environment",
   "../service!pentaho.config.IService?single",
   "./changes/Transaction",
   "./changes/TransactionScope",
   "./changes/CommittedScope",
   "../lang/Base",
+  "./util",
   "../util/promise",
   "../util/arg",
   "../util/error",
   "../util/object",
-  "../util/fun"
-], function(localRequire, module, service, typeInfo, bundle, standard, SpecificationContext, SpecificationScope,
-    mainPlatformEnv, configurationService,
+  "../util/fun",
+  "../debug",
+  "../debug/Levels",
+  "../util/logger",
+  "./standard" // so that r.js sees otherwise invisible dependencies.
+], function(localRequire, module, typeInfo, bundle, InstanceConstructionContext,
+    SpecificationContext, SpecificationScope,
+    InstancesContainer, mainPlatformEnv, configurationService,
     Transaction, TransactionScope, CommittedScope,
-    Base, promiseUtil, arg, error, O, F) {
+    Base, typeUtil, promiseUtil, arg, error, O, F, debugMgr, DebugLevels, logger) {
 
   "use strict";
 
   /* globals Promise */
 
-  var __nextFactoryUid = 1;
-  var __singleton = null;
+  /* eslint dot-notation: 0, no-unexpected-multiline: 0 */
 
   // Default `base` type in a type specification.
   var __defaultBaseTypeMid = "complex";
 
+  var __instanceTypeId = "pentaho/type/instance";
+
   var O_hasOwn = Object.prototype.hasOwnProperty;
 
+  var __standardIds = Object.freeze([
+    // types
+    "instance",
+    "value",
+    "element",
+    "list",
+    "simple",
+    "string",
+    "number",
+    "boolean",
+    "date",
+    "complex",
+    "object",
+    "function",
+    "property",
+    "model",
+    "application",
+    "mixins/enum"
+  ].map(function(id) { return "pentaho/type/" + id; }));
+
+  /**
+   * @name pentaho.type.ITypeHolder
+   * @interface
+   * @private
+   *
+   * @property {string} id - The type's identifier. Always set.
+   * @property {?function} factory - The type's factory function. Set when the AMD module is loaded.
+   * @property {!Array.<string>} deps - The type's dependencies. Set when the AMD module is loaded.
+   *
+   * @property {Class.<pentaho.type.Instance>} Ctor - The type's constructor.
+   * Only not `null` when the type has already been created.
+   *
+   * @property {Promise.<Class.<pentaho.type.Instance>>} promise - A promise for the type's constructor.
+   * Only not `null` since the type has begun loading.
+   */
+
+  var TypeHolder = Base.extend({
```

Comment:
Isn't `TypeHolder` similar to `InstancesContainer`? Maybe move it to its own file?

Reply:
Ok, was seeing it wrong... `InstanceHolder` is a inner class of `InstancesContainer`, as `TypeHolder` is a inner class of `Context`.

Reply:
I believe that you mean that `TypeHolder` is similar to `InstanceHolder`. Yes they are very similar, but are also different in several places. Creating a separate base class would be an additional effort for me, specially taking into account the time it already took to develop all this. On a second stride, we would do that abstraction.